File: src/test/resources/de/hhu/stups/codegenerator/project4/BTuple.cpp
#include <iostream>
#include <string>
#include "BBoolean.cpp"

using namespace std;

#ifndef BTUPLE_H
#define BTUPLE_H

template<typename S, typename T>
class BTuple : public BObject {

	private:
	    S lhs;
	    T rhs;

	public:
	    typedef void value_type;
        typedef S left_type;
        typedef T right_type;

        BTuple(const S& l, const T& r) {
            lhs = l;
            rhs = r;
        }

        BTuple(){}

        S projection1() {
            return lhs;
        }

        T projection2() {
            return rhs;
        }

        BBoolean equal(const BTuple& other) {
            return lhs == other.lhs && rhs == other.rhs;
        }

        BBoolean unequal(const BTuple& other) {
            return lhs != other.lhs || rhs != other.rhs;
        }

        friend bool operator ==(const BTuple<S,T>& o1, const BTuple<S,T>& o2) {
            return o1.lhs == o2.lhs && o1.rhs == o2.rhs;
        }

        friend bool operator !=(const BTuple<S,T>& o1, const BTuple<S,T>& o2) {
            return o1.lhs != o2.lhs || o1.rhs != o2.rhs;
        }

        void operator =(const BTuple<S,T>& other) {
            this->lhs = other.lhs;
            this->rhs = other.rhs;
        }

        int hashCode() const {
            return this->lhs.hashCode() ^ (this->rhs.hashCode() << 1);
        }

        friend std::ostream& operator<<(std::ostream &strm, const BTuple<S,T> &b) {
          return strm << "(" << b.lhs << " |-> " << b.rhs << ")";
        }

        /*BTuple operator() {
            return *this;
        }*/

	/*public boolean equals(Object o) {
		if (this == o) {
			return true;
		}
		if (o == null || getClass() != o.getClass()) {
			return false;
		}

        BTuple bObjects = (BTuple) o;
		// elements is never null
		return bObjects.getFirst().equals(this.first) && bObjects.getSecond().equals(this.second);
	}

	public int hashCode() {
		return Objects.hash(first, second);
	}

	@Override
	public java.lang.String toString() {
		return "(" + this.getFirst() + " |-> " + this.getSecond() + ')';
	}*/

};
#endif

File: btypes_primitives/src/main/python_pyrsistant_immutable/btypes/BTuple.py
from btypes.BObject import *
from btypes.BBoolean import *


class BTuple(BObject):

    def __init__(self, first, second):
        if first is None or second is None:
            raise ValueError()
        self.first = first
        self.second = second

    def __eq__(self, other):
        return self.equals(other)

    def equals(self, other) -> 'bool':
        if self is other:
            return True
        if other is None or type(self) != type(other):
            return False

        b_objects = other
        # elements is never None
        return b_objects.projection1().__eq__(self.first) and b_objects.projection2().__eq__(self.second)

    def __hash__(self):
        return hash(hash(self.first) + hash(self.second))

    def projection1(self):
        return self.first

    def projection2(self):
        return self.second

    def __str__(self):
        return "(" + str(self.projection1()) + " |-> " + str(self.projection2()) + ')'

    def equal(self, other: 'BTuple') -> 'BBoolean':
        return BBoolean(self.equals(other))

    def unequal(self, other: 'BTuple') -> 'BBoolean':
        return BBoolean(not self.equals(other))
